### DISCUSSION TITLE
[React Native] Atom updates that occur within setTimeout/setInterval trigger a re-render for every useAtom instance

### DISCUSSION THREAD

**Discussion by computerjazz** [pmndrs/jotai#2287] [Bug report]
## Summary
In React Native, atom updates that occur within a setTimeout trigger a re-render _for each instance_ of the `useAtom` hook present (normally the multiple other hook instances would probably be buried within other hooks, but for the purposes of the demo I added a bunch to the main Screen). 

Notice that the "inline" update causes one re-render (expected), while the update that occurs within the setTimeout causes 10 re-renders (one for each instance of `useAtom`, not expected)

(note that the same code running in a browser does NOT have the issue: https://codesandbox.io/p/sandbox/condescending-sunset-p4dcx4?file=%2Fsrc%2FApp.js)

## Link to reproduction

https://snack.expo.dev/@easydan/jotai-re-render-bug

## Check List

Please do not ask questions in issues.

- [ ] I've already opened a [discussion](https://github.com/pmndrs/jotai/discussions) before opening this issue, or already discussed in other media.

Please include a minimal reproduction.

- [x] I've added a link to a [typescript playground](https://www.typescriptlang.org/play) or [codesandbox](https://codesandbox.io) with a minimal reproduction.

**Comment by dai-shi** [comment]
Converting to a discussion for now. We'll open a new issue after investigating it.

**Comment by dai-shi** [comment]
Thanks for reporting.
How can I see the console logs in the snack?

**Reply by computerjazz** [reply]
There's a Logs panel at the bottom (you might have to be running in a sim before it shows up)
![Screenshot 2023-11-30 at 5 57 22 PM](https://github.com/pmndrs/jotai/assets/6730148/1beb5d24-bd14-45ce-bafc-9da44b7a0214)

**Reply by dai-shi** [reply]
Oh, thanks. So, it doesn't reproduce on react-native-web:
<img width="401" alt="image" src="https://github.com/pmndrs/jotai/assets/490574/0bbc0c95-045d-4b40-817e-73f0761f97d5">

**Reply by dai-shi** [reply]
And, it does on iPhone.
<img width="372" alt="image" src="https://github.com/pmndrs/jotai/assets/490574/7bb69615-1cf3-4a13-83bb-20365cc1f11f">

**Reply by dai-shi** [reply]
Changing the code to this fixes it.
```js
  const update = () => unstable_batchedUpdates(() => {
    setVal(Math.random())
  })
```
<img width="353" alt="image" src="https://github.com/pmndrs/jotai/assets/490574/b3b616eb-ddff-4bf1-a803-e02d1d494e4b">

So, I'm pretty sure this is causing because automatic batching isn't effective in RN or Expo.

There's nothing Jotai library will do, and it's a valid workaround to use `unstable_batchedUpdates` in such a case, like React 17.

**Reply by computerjazz** [reply]
Found this

> Note: Users on React Native 0.69, but still on the Old Architecture will still use React 17 mode even if React 18 is installed in the package.json file.

https://reactnative.dev/docs/react-18-and-react-native

**Reply by computerjazz** [reply]
And thank you! `unstable_batchedUpdates` solved my performance issues

**Comment by dai-shi** [comment]
https://codesandbox.io/p/sandbox/condescending-sunset-p4dcx4?file=%2Fsrc%2FApp.js
I wonder if changing it to React 17 will produce the same result like the expo snack.

**Reply by computerjazz** [reply]
worth a try, but fwiw the react native project I'm working on (not the snack) is on `    "react": "^18.2.0",`

**Reply by computerjazz** [reply]
React 17 does display some extra re-renders, although not all 10: https://codesandbox.io/p/sandbox/react-17-forked-ymhvk7?file=%2Fsrc%2FApp.js%3A28%2C5
![Screenshot 2023-12-01 at 2 12 32 PM](https://github.com/pmndrs/jotai/assets/6730148/0309f850-8b99-49a5-800c-beb1efc265bd)
